### DISCUSSION TITLE
CosineSimilarity code Explain

### DISCUSSION THREAD

**Discussion by mtsol** [NVIDIA/spark-rapids#7790] [Q&A]
Hello team, I need to know the function call at line 34 in https://github.com/NVIDIA/spark-rapids-examples/blob/branch-22.12/examples/UDF-Examples/RAPIDS-accelerated-UDFs/src/main/java/com/nvidia/spark/rapids/udf/java/CosineSimilarity.java executes on GPU or CPU? And if it executes on CPU then where is the GPU code?

**Comment by jlowe** [comment]
Line 34 provides the CPU implementation of the UDF.  Per the [RAPIDS accelerated UDF documentation](https://github.com/NVIDIA/spark-rapids/blob/branch-23.04/docs/additional-functionality/rapids-udfs.md#adding-gpu-implementations-to-udfs) the UDF provides _two_ implementations, the existing one for the CPU and a new one for the GPU.  The GPU implementation of the UDF is provided by the `evaluateColumnar` interface, which starts at line 64 in that same file.  This method ensures the native library is loaded and then proceeds to evaluate it on the GPU via a JNI method.

The JNI method source is at https://github.com/NVIDIA/spark-rapids-examples/blob/branch-22.12/examples/UDF-Examples/RAPIDS-accelerated-UDFs/src/main/cpp/src/CosineSimilarityJni.cpp#L60-L92 and the source of the GPU code invoked by that JNI method is at https://github.com/NVIDIA/spark-rapids-examples/blob/branch-22.12/examples/UDF-Examples/RAPIDS-accelerated-UDFs/src/main/cpp/src/cosine_similarity.cu.
